File: scripts/analytics.py
"""조회조건 → 대시보드 뷰모델 계산. 대시보드와 PDF 리포트가 공용으로 사용한다.

granularity: 'day' | 'month' | 'year'
  - day   : period=='daily' 레코드만, 버킷=날짜(YYYY-MM-DD)
  - month : 모든 레코드, 버킷=YYYY-MM (일별은 그 달로 합산)
  - year  : 모든 레코드, 버킷=YYYY
start/end 는 granularity 와 같은 포맷의 문자열.
"""
from __future__ import annotations

import calendar
import sys
from collections import OrderedDict
from datetime import date, datetime
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
import store

# 채널 구분용 팔레트 (색상 대비 큰 6~13색)
PALETTE = ["#0067ac", "#ef8a2b", "#3aa757", "#d64550", "#8a63c9", "#00a1b5",
           "#c9a227", "#5b7089", "#e0748a", "#2f8f8f", "#7a6fbe", "#b0611f", "#4c78a8"]

GRAN_LABEL = {
    "day":   {"unit": "일", "latest": "금일 유입",     "delta": "전일 대비", "col": "최근 이용건"},
    "month": {"unit": "월", "latest": "해당 월 이용건", "delta": "전월 대비", "col": "월 이용건"},
    "year":  {"unit": "년", "latest": "해당 년 이용건", "delta": "전년 대비", "col": "연 이용건"},
}


def _pct(cur: float, prev: float) -> float | None:
    if prev == 0:
        return None
    return round((cur - prev) / prev * 100, 1)


def visible_channels(master: dict, *, for_display: bool = True) -> list[dict]:
    chans = master.get("channels", [])
    if for_display:
        chans = [c for c in chans if c.get("display_yn", "Y") == "Y"]
    return sorted(chans, key=lambda c: c.get("sort", 999))


def _bucket_of(date_str: str, gran: str) -> str:
    if gran == "year":
        return date_str[:4]
    if gran == "month":
        return date_str[:7]
    return date_str


def build_view(*, granularity: str = "day", start: str | None = None,
               end: str | None = None, channels: list[str] | None = None,
               **_ignore) -> dict:
    gran = granularity if granularity in GRAN_LABEL else "day"
    inflow = store.load_inflow()
    master = store.load_channels()
    log = store.load_log()

    records = sorted(inflow.get("records", []), key=lambda r: r["date"])
    all_dates = [r["date"] for r in records]

    disp = visible_channels(master, for_display=True)
    order = {c["code"]: i for i, c in enumerate(disp)}
    name_by = {c["code"]: c.get("name", c["code"]) for c in disp}
    type_by = {c["code"]: c.get("type", "") for c in disp}
    color_by = {c["code"]: PALETTE[i % len(PALETTE)] for i, c in enumerate(disp)}
    allowed = set(name_by)
    if channels:
        allowed &= set(channels)

    # day 모드는 일별 레코드만, 나머지는 전체
    pool = [r for r in records if r.get("period", "daily") == "daily"] if gran == "day" else records
    pool_buckets = sorted({_bucket_of(r["date"], gran) for r in pool})

    if not start:
        start = pool_buckets[0] if pool_buckets else _bucket_of(date.today().isoformat(), gran)
    if not end:
        end = pool_buckets[-1] if pool_buckets else _bucket_of(date.today().isoformat(), gran)
    if start > end:
        start, end = end, start

    in_range = [r for r in pool if start <= _bucket_of(r["date"], gran) <= end]

    # 버킷 → 채널 → {y,n,total}
    agg: "OrderedDict[str, dict]" = OrderedDict()
    for r in in_range:
        b = _bucket_of(r["date"], gran)
        slot = agg.setdefault(b, {})
        for c in r.get("channels", []):
            if c["channel"] not in allowed:
                continue
            cs = slot.setdefault(c["channel"], {"y": 0, "n": 0, "total": 0})
            cs["y"] += c["y"]; cs["n"] += c["n"]; cs["total"] += c["total"]
    buckets = sorted(agg)

    # 추이 / 누적
    trend, cumulative = [], []
    run = 0
    for b in buckets:
        rows = agg[b].values()
        t = sum(x["total"] for x in rows)
        y = sum(x["y"] for x in rows)
        n = sum(x["n"] for x in rows)
        run += t
        trend.append({"date": b, "total": t, "y": y, "n": n})
        cumulative.append({"date": b, "total": run})

    period_total = {k: sum(d[k] for d in trend) for k in ("total", "y", "n")}

    # 최신 버킷 / 직전 버킷
    latest = trend[-1] if trend else {"date": None, "total": 0, "y": 0, "n": 0}

    # 최신 버킷이 '진행 중(부분 집계)' 인지 — 월/년의 마지막 데이터가 그 구간 말일이 아닐 때
    latest_partial = False
    if trend and all_dates and gran != "day":
        dmax = all_dates[-1]
        yy, mm, dd = (int(x) for x in dmax.split("-"))
        if gran == "month":
            latest_partial = _bucket_of(dmax, "month") == latest["date"] and dd < calendar.monthrange(yy, mm)[1]
        else:  # year
            latest_partial = dmax[:4] == latest["date"] and (mm, dd) < (12, 31)

    delta = {"prev_date": None, "total_pct": None, "total_diff": 0, "note": None}
    if len(trend) >= 2:
        prev = trend[-2]
        if latest_partial:
            delta["note"] = f"최신 {GRAN_LABEL[gran]['unit']}({latest['date']})은 아직 진행 중이라 비교 생략"
        else:
            delta = {
                "prev_date": prev["date"],
                "total_pct": _pct(latest["total"], prev["total"]),
                "total_diff": latest["total"] - prev["total"],
                "note": None,
            }
    elif trend:
        delta["note"] = "비교할 이전 구간 없음"

    # 채널별
    last_b = buckets[-1] if buckets else None
    prev_b = None if latest_partial else (buckets[-2] if len(buckets) >= 2 else None)
    by_channel = []
    for code in sorted(allowed, key=lambda c: order.get(c, 999)):
        series = [{"date": b, "total": agg[b].get(code, {}).get("total", 0)} for b in buckets]
        p_total = sum(agg[b].get(code, {}).get("total", 0) for b in buckets)
        p_y = sum(agg[b].get(code, {}).get("y", 0) for b in buckets)
        p_n = sum(agg[b].get(code, {}).get("n", 0) for b in buckets)
        cur = agg.get(last_b, {}).get(code) if last_b else None
        prv = agg.get(prev_b, {}).get(code) if prev_b else None
        by_channel.append({
            "channel": code,
            "name": name_by.get(code, code),
            "type": type_by.get(code, ""),
            "color": color_by.get(code, "#888"),
            "today_total": (cur or {}).get("total", 0),
            "today_y": (cur or {}).get("y", 0),
            "today_n": (cur or {}).get("n", 0),
            "period_total": p_total,
            "period_y": p_y,
            "period_n": p_n,
            "share_pct": round(p_total / period_total["total"] * 100, 1) if period_total["total"] else 0.0,
            "dod_pct": _pct(cur["total"], prv["total"]) if (cur and prv) else None,
            "series": series,
        })
    by_channel.sort(key=lambda c: c["period_total"], reverse=True)

    return {
        "query": {
            "granularity": gran, "start": start, "end": end,
            "channels": sorted(channels) if channels else [],
        },
        "labels": GRAN_LABEL[gran],
        "buckets": buckets,
        "as_of": {
            "latest_date": all_dates[-1] if all_dates else None,
            "record_count": len(records),
            "last_collected_at": (log.get("logs") or [{}])[-1].get("collected_at"),
            "date_min": all_dates[0] if all_dates else None,
            "date_max": all_dates[-1] if all_dates else None,
        },
        "kpi": {
            "today": latest,
            "latest_partial": latest_partial,
            "period_total": period_total,
            "dod": delta,
            "active_channels": len(disp),
            "channels_in_view": len(allowed),
        },
        "daily_trend": trend,
        "cumulative_trend": cumulative,
        "by_channel": by_channel,
        "channel_share": [
            {"name": c["name"], "total": c["period_total"], "pct": c["share_pct"], "color": c["color"]}
            for c in by_channel if c["period_total"] > 0
        ],
        "channel_master": [
            {**c, "color": color_by.get(c["code"], "#888")} for c in disp
        ],
        "available_dates": all_dates,
    }


if __name__ == "__main__":
    import json
    for g in ("day", "month", "year"):
        v = build_view(granularity=g)
        print(f"\n=== {g} === buckets={v['buckets']}")
        print("  kpi.today:", v["kpi"]["today"])
        print("  by_channel:", [(c["name"], c["period_total"]) for c in v["by_channel"]])

File: docs/analytics.js
/* 조회조건 → 대시보드 뷰모델 계산 (scripts/analytics.py 의 클라이언트 포팅).
   GitHub Pages(정적 호스팅)에서 data/*.json 만으로 동작한다.
   window.ANALYTICS.buildView({granularity, start, end, channels}, {inflow, channels: master, log}) */
(function () {
  "use strict";

  const PALETTE = ["#0067ac", "#ef8a2b", "#3aa757", "#d64550", "#8a63c9", "#00a1b5",
    "#c9a227", "#5b7089", "#e0748a", "#2f8f8f", "#7a6fbe", "#b0611f", "#4c78a8"];

  const GRAN_LABEL = {
    day:   { unit: "일", latest: "금일 유입",     delta: "전일 대비", col: "최근 이용건" },
    month: { unit: "월", latest: "해당 월 이용건", delta: "전월 대비", col: "월 이용건" },
    year:  { unit: "년", latest: "해당 년 이용건", delta: "전년 대비", col: "연 이용건" },
  };

  const round1 = (n) => Math.round(n * 10) / 10;
  function pct(cur, prev) { return prev === 0 ? null : round1(((cur - prev) / prev) * 100); }
  function daysInMonth(y, m) { return new Date(y, m, 0).getDate(); }
  function bucketOf(dateStr, gran) {
    return gran === "year" ? dateStr.slice(0, 4) : gran === "month" ? dateStr.slice(0, 7) : dateStr;
  }
  function todayISO() {
    const d = new Date();
    return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}-${String(d.getDate()).padStart(2, "0")}`;
  }

  function visibleChannels(master) {
    return (master.channels || [])
      .filter((c) => (c.display_yn || "Y") === "Y")
      .slice()
      .sort((a, b) => (a.sort ?? 999) - (b.sort ?? 999));
  }

  function buildView(q, data) {
    const gran = GRAN_LABEL[q.granularity] ? q.granularity : "day";
    const inflow = data.inflow || { records: [] };
    const master = data.channels || { channels: [] };
    const log = data.log || { logs: [] };

    const records = (inflow.records || []).slice().sort((a, b) => (a.date < b.date ? -1 : a.date > b.date ? 1 : 0));
    const allDates = records.map((r) => r.date);

    const disp = visibleChannels(master);
    const order = {}; disp.forEach((c, i) => (order[c.code] = i));
    const nameBy = {}, typeBy = {}, colorBy = {};
    disp.forEach((c, i) => { nameBy[c.code] = c.name || c.code; typeBy[c.code] = c.type || ""; colorBy[c.code] = PALETTE[i % PALETTE.length]; });

    let allowed = new Set(disp.map((c) => c.code));
    if (q.channels && q.channels.length) allowed = new Set(q.channels.filter((c) => allowed.has(c)));

    const pool = gran === "day" ? records.filter((r) => (r.period || "daily") === "daily") : records;
    const poolBuckets = [...new Set(pool.map((r) => bucketOf(r.date, gran)))].sort();

    let start = q.start || (poolBuckets[0] || bucketOf(todayISO(), gran));
    let end = q.end || (poolBuckets[poolBuckets.length - 1] || bucketOf(todayISO(), gran));
    if (start > end) { const t = start; start = end; end = t; }

    const inRange = pool.filter((r) => { const b = bucketOf(r.date, gran); return b >= start && b <= end; });

    const agg = {};
    for (const r of inRange) {
      const b = bucketOf(r.date, gran);
      const slot = (agg[b] = agg[b] || {});
      for (const c of r.channels || []) {
        if (!allowed.has(c.channel)) continue;
        const cs = (slot[c.channel] = slot[c.channel] || { y: 0, n: 0, total: 0 });
        cs.y += c.y; cs.n += c.n; cs.total += c.total;
      }
    }
    const buckets = Object.keys(agg).sort();

    const trend = [], cumulative = [];
    let run = 0;
    for (const b of buckets) {
      const rows = Object.values(agg[b]);
      const t = rows.reduce((s, x) => s + x.total, 0);
      const y = rows.reduce((s, x) => s + x.y, 0);
      const n = rows.reduce((s, x) => s + x.n, 0);
      run += t;
      trend.push({ date: b, total: t, y, n });
      cumulative.push({ date: b, total: run });
    }
    const period_total = { total: 0, y: 0, n: 0 };
    trend.forEach((d) => { period_total.total += d.total; period_total.y += d.y; period_total.n += d.n; });

    const latest = trend.length ? trend[trend.length - 1] : { date: null, total: 0, y: 0, n: 0 };

    let latest_partial = false;
    if (trend.length && allDates.length && gran !== "day") {
      const dmax = allDates[allDates.length - 1];
      const [yy, mm, dd] = dmax.split("-").map(Number);
      if (gran === "month") latest_partial = bucketOf(dmax, "month") === latest.date && dd < daysInMonth(yy, mm);
      else latest_partial = dmax.slice(0, 4) === latest.date && (mm < 12 || dd < 31);
    }

    let delta = { prev_date: null, total_pct: null, total_diff: 0, note: null };
    if (trend.length >= 2) {
      const prev = trend[trend.length - 2];
      if (latest_partial) delta.note = `최신 ${GRAN_LABEL[gran].unit}(${latest.date})은 아직 진행 중이라 비교 생략`;
      else delta = { prev_date: prev.date, total_pct: pct(latest.total, prev.total), total_diff: latest.total - prev.total, note: null };
    } else if (trend.length) delta.note = "비교할 이전 구간 없음";

    const lastB = buckets[buckets.length - 1] || null;
    const prevB = latest_partial ? null : (buckets.length >= 2 ? buckets[buckets.length - 2] : null);

    let by_channel = [...allowed].sort((a, b) => (order[a] ?? 999) - (order[b] ?? 999)).map((code) => {
      const series = buckets.map((b) => ({ date: b, total: (agg[b][code] || {}).total || 0 }));
      let p_total = 0, p_y = 0, p_n = 0;
      buckets.forEach((b) => { const x = agg[b][code]; if (x) { p_total += x.total; p_y += x.y; p_n += x.n; } });
      const cur = lastB ? agg[lastB][code] : null;
      const prv = prevB ? agg[prevB][code] : null;
      return {
        channel: code, name: nameBy[code] || code, type: typeBy[code] || "", color: colorBy[code] || "#888",
        today_total: (cur || {}).total || 0, today_y: (cur || {}).y || 0, today_n: (cur || {}).n || 0,
        period_total: p_total, period_y: p_y, period_n: p_n,
        share_pct: period_total.total ? round1((p_total / period_total.total) * 100) : 0,
        dod_pct: cur && prv ? pct(cur.total, prv.total) : null,
        series,
      };
    });
    by_channel.sort((a, b) => b.period_total - a.period_total);

    const lastLog = (log.logs && log.logs.length) ? log.logs[log.logs.length - 1] : {};

    return {
      query: { granularity: gran, start, end, channels: (q.channels || []).slice().sort() },
      labels: GRAN_LABEL[gran],
      buckets,
      as_of: {
        latest_date: allDates[allDates.length - 1] || null,
        record_count: records.length,
        last_collected_at: lastLog.collected_at || null,
        date_min: allDates[0] || null,
        date_max: allDates[allDates.length - 1] || null,
      },
      kpi: {
        today: latest, latest_partial, period_total, dod: delta,
        active_channels: disp.length, channels_in_view: allowed.size,
      },
      daily_trend: trend,
      cumulative_trend: cumulative,
      by_channel,
      channel_share: by_channel.filter((c) => c.period_total > 0)
        .map((c) => ({ name: c.name, total: c.period_total, pct: c.share_pct, color: c.color })),
      channel_master: disp.map((c, i) => Object.assign({}, c, { color: PALETTE[i % PALETTE.length] })),
      available_dates: allDates,
    };
  }

  window.ANALYTICS = { buildView, PALETTE, GRAN_LABEL };
})();
